utils/backoff.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"time"
+)
+
+const (
+	backoffBase uint64 = 128
+	maxAttempts        = 4
+)
+
+// BackoffOperation is executed by Retry.
+// The BackoffOperation will be retried the provided
+// number attempts if returning an error.
+type BackoffOperation func() error
+
+// ProgressiveRetry executes a BackoffOperation and retries the operation 3 times upon error.
+func ProgressiveRetry(bo BackoffOperation) error {
```

Comment:
1/5 would suggest adding an `retryErrorf func(error) bool` parameter that would decide if the error is to retried (true if nil), it can also log if needed.

I would also 0/5 consider using a multierror wrapper to return them all.

Both can be a separate PR though.

Reply:
I can see the value of a `retryErrorf`, but at that point I'd almost ask if we need this whole helper function at all, and perhaps the calling code should just `for attempt := 1; attempt <= 3; attempt++ { ... }` and do any custom handling it requires there. There isn't a /lot/ of helper code here to begin with.